src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs

```diff
@@ -527,16 +527,11 @@ private void Visit(IList<Notification> notifications, string parentPointer, stri
 
         private void Visit(Resources resources, string resourcesPointer)
         {
-            Rule[] rules = resources.Rules.Values.ToArray();
             string rulesPointer = resourcesPointer.AtProperty(SarifPropertyName.Rules);
 
-            for (int i = 0; i < rules.Length; ++i)
+            foreach (string key in resources.Rules.Keys)
             {
-                Rule rule = rules[i];
-                if (rule.Id != null)
-                {
-                    Visit(rule, rulesPointer.AtProperty(rule.Id));
-                }
+                Visit(resources.Rules[key], rulesPointer.AtProperty(key));
```

Comment:
Yes. Our current VisitFileData pattern provides a file data instance that might not itself contain property data that is expressed as the key used to store the file data instance in the files table. this lessens there is currently no easy way to override the visitor to try to persist this information as instance data available when the callback is executed, a serious design problem.

---
In reply to: [214738173](https://github.com/Microsoft/sarif-sdk/pull/1068#discussion_r214738173) [](ancestors = 214738173,214738098)

Reply:
That will be difficult to generalize because the "matching" properties are so different in those three cases:
- For the `rules` dictionary, the matching property is `rule.id`.
- For the `files` dictionary, the matching property is `file.fileLocation.uri` -- not even a direct child of the file object.
- For the logical locations dictionary, there are in fact _two_properties (`name` and `fullyQualifiedName`) that might match the dictionary key.
So as desirable as it would be to treat these cases uniformly, I think there needs to be separate code for each of those dictionaries.

---
In reply to: [214738900](https://github.com/Microsoft/sarif-sdk/pull/1068#discussion_r214738900) [](ancestors = 214738900,214738173,214738098)

Reply:
The uniform approach is to provide the string-valued key for any item that is visited because it is stored in a SARIF dictionary. The visited override must understand the semantics of the key value.

---
In reply to: [214739878](https://github.com/Microsoft/sarif-sdk/pull/1068#discussion_r214739878) [](ancestors = 214739878,214738900,214738173,214738098)

Reply:
@michaelcfanning As you pointed out in email: the generalization is to pass the dictionary key to the visitors for `files`, `logicalLocations`, and `rules`.

---
In reply to: [214739878](https://github.com/Microsoft/sarif-sdk/pull/1068#discussion_r214739878) [](ancestors = 214739878,214738900,214738173,214738098)